# 简介
Cactus是一个使用progressive策略进行多重全基因组比对工具。简单的说，对于一组已知系统发育关系的基因组，将最近缘的两物种互相比对，构建祖先序列，重复这个过程直至构建出整个树的共同祖先。这个策略的优势在于不受参考基因组限制， 同时也能节省计算资源。劣势在于在远源物种的比对实际上是构建出的祖先基因组比对而来，如果构建祖先基因时出问题，会影响所有上游比对的的效果。
# 安装
安装可以选择docker镜像或者是官方repositry下载二进制包。我一般选择下载二进制包，地址在这里https://github.com/ComparativeGenomicsToolkit/cactus/releases  
安装过程可以参考release的BIN-INSTALL.md,我习惯使用conda 构建虚拟环境进行:
```
wget https://github.com/ComparativeGenomicsToolkit/cactus/releases/download/v2.9.2/cactus-bin-v2.9.2.tar.gz
tar -xzf cactus-bin-v2.9.2.tar.gz
cd cactus-bin-v2.9.2
conda create -n cactus pip
pip install -U setuptools pip wheel
pip install -U .
pip install -U -r ./toil-requirement.txt
```
这里主体就安装完成了。一些工具需要使用UCSC kent tools，可以使用bioconda安装，用可以从UCSC下载:
```
cd bin && for i in wigToBigWig faToTwoBit bedToBigBed bigBedToBed axtChain pslPosTarget bedSort hgGcPercent mafToBigMaf hgLoadMafSummary hgLoadChain; do wget -q http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/${i}; chmod +x ${i}; done
```


# 输入文件
Cactus的输入文件主要有两部分，一个是基因组，另一个是带枝长的物种树。将树和基因组名字和路径写入一个输入文件中即可。这里我们命名为"cactus.input":
```
((camel:0.024449399999999996,(pig:0.026929899999999993,((xilu:0.0318519,((pronghorn:0.011900899999999999,(giraffe:0.004092180000000001,okapi:0.006236209999999999):0.004874759999999999):7.061509999999986E-4,((muskdeer:0.009617210000000001,(cattle:0.007105599999999997,(sheep:0.0023832800000000015,goat:0.0023733999999999977):0.0058297):0.0019783300000000004):9.002019999999993E-4,(((reev_muntjac:0.0041239,ECEP:0.003908189999999999):9.83137000000002E-4,(TUNLU:0.0027520899999999987,(TL:0.002494690000000001,((mhl:7.166099999999995E-4,ML:7.621969999999992E-4):9.152210000000008E-4,PL:0.0019297900000000007):4.7706499999999874E-4):2.276429999999996E-4):0.0012576899999999988):9.57613000000003E-4,((HJ:0.0041008699999999995,PZ:0.0041008699999999995):0.002634530000000003,(mule_deer:0.0028536499999999992,XL:0.00281675):0.0021085500000000007):9.272520000000034E-4):0.005689299999999998):8.850460000000004E-4):0.011183300000000004):0.010619599999999993,(hippo:0.0192975,(gray_whale:0.007563460000000001,killer_whale:0.009708890000000005):0.010820759999999999):0.002187310000000005):0.004391329999999999):0.0033151899999999956):0.0162236,tapir:0.0162236);
camel /public3/home/a6s000219/lizihe/01.data/01.genome/changename/camel.fa
cattle /public3/home/a6s000219/lizihe/01.data/01.genome/changename/cattle.fa
ECEP /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/ECEP.fa
giraffe /public3/home/a6s000219/lizihe/01.data/01.genome/changename/giraffe.fa
goat /public3/home/a6s000219/lizihe/01.data/01.genome/changename/goat.fa
gray_whale /public3/home/a6s000219/lizihe/01.data/01.genome/changename/gray_whale.fa
hippo /public3/home/a6s000219/lizihe/01.data/01.genome/changename/hippo.fa
HJ /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/HJ.fa
killer_whale /public3/home/a6s000219/lizihe/01.data/01.genome/changename/killer_whale.fa
mhl /public3/home/a6s000219/lizihe/01.data/01.genome/changename/mhl.fa
ML /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/ML.fa
mule_deer /public3/home/a6s000219/lizihe/01.data/01.genome/changename/mule_deer.fa
muskdeer /public3/home/a6s000219/lizihe/01.data/01.genome/changename/muskdeer.fa
okapi /public3/home/a6s000219/lizihe/01.data/01.genome/changename/okapi.fa
pig /public3/home/a6s000219/lizihe/01.data/01.genome/changename/pig.fa
PL /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/PL.fa
pronghorn /public3/home/a6s000219/lizihe/01.data/01.genome/changename/pronghorn.fa
PZ /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/PZ.fa
reev_muntjac /public3/home/a6s000219/lizihe/01.data/01.genome/changename/reev_muntjac.fa
sheep /public3/home/a6s000219/lizihe/01.data/01.genome/changename/sheep.fa
tapir /public3/home/a6s000219/lizihe/01.data/01.genome/changename/tapir.fa
TL /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/TL.fa
TUNLU /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/TUNLU.fa
xilu /public3/home/a6s000219/lizihe/01.data/01.genome/changename/xilu.fa
XL /public3/home/a6s000219/lizihe/01.data/01.genome/luke_mask/XL.fa
```
# 运行Cactus
运行Cactus主要有三种选择，一种是直接使用内置的TOIL任务管理器运行，适用于小规模数据集在本地跑或者SLURM集群（v2.61以后，未测试）。  
第二种是生成WDL，使用cromwell等任务管理器运行，可以适用于大部分集群系统。  
第三种类似与第二种，但是生成的是每一步的bash指令，需要你自己运行管理，反过来说也最灵活，可以在不同的地方跑，最后结果合并到一起。
## Cromwell方法
待施工
## Step by step 方法
直接运行cactus-prepare:
```
cactus-prepare cactus.input  --outDir steps-output --outSeqFile steps-output/evovlerMammals.txt --outHal steps-output/evolverMammals.hal --jobStore jobstore --preprocessBatchSize 1 > cactus_run_all.sh
```
输出的文件如下，每一个##是一个块，块与块之间都是串行的，运行完一个块里的所有命令，才可以运行下一个。  
在块内部，Preprocessor每一行相互独立，可以并行运行；Round X 每三行一个部分相互独立，可以并行；HAL Merging只能串行。  
你可以看到，除了最开始的preprocess和最后的 HAL merging，中间的过程都是在重复“cactus-blast --> cactus-align --> hal2fasta”这个两两比对生成祖先序列的过程。  
其中cactus-blast是可以多核的，cactus-align是最耗费内存的一步，对于两个3g的基因组，预估内存需要500GB。
```
## generated by : /public3/home/a6s000219/libin/software/miniconda3/envs/cactus/bin/cactus-prepare cactus.input --outDir steps-output --outSeqFile steps-output/evovlerMammals.txt --outHal steps-output/evolverMammals.hal --jobStore jobstore --preprocessBatchSize 1
## date : 2024-10-28 15:28:22.902986
## cactus commit : 7286b49b264896f43cc64aa405b39f914d43f75b


## Preprocessor
cactus-preprocess jobstore/0 cactus.input steps-output/evovlerMammals.txt --inputNames tapir   --logFile steps-output/logs/preprocess-tapir.log
cactus-preprocess jobstore/1 cactus.input steps-output/evovlerMammals.txt --inputNames camel   --logFile steps-output/logs/preprocess-camel.log
cactus-preprocess jobstore/2 cactus.input steps-output/evovlerMammals.txt --inputNames pig   --logFile steps-output/logs/preprocess-pig.log
cactus-preprocess jobstore/3 cactus.input steps-output/evovlerMammals.txt --inputNames xilu   --logFile steps-output/logs/preprocess-xilu.log
cactus-preprocess jobstore/4 cactus.input steps-output/evovlerMammals.txt --inputNames hippo   --logFile steps-output/logs/preprocess-hippo.log
cactus-preprocess jobstore/5 cactus.input steps-output/evovlerMammals.txt --inputNames gray_whale   --logFile steps-output/logs/preprocess-gray_whale.log
cactus-preprocess jobstore/6 cactus.input steps-output/evovlerMammals.txt --inputNames killer_whale   --logFile steps-output/logs/preprocess-killer_whale.log
cactus-preprocess jobstore/7 cactus.input steps-output/evovlerMammals.txt --inputNames pronghorn   --logFile steps-output/logs/preprocess-pronghorn.log
cactus-preprocess jobstore/8 cactus.input steps-output/evovlerMammals.txt --inputNames giraffe   --logFile steps-output/logs/preprocess-giraffe.log
cactus-preprocess jobstore/9 cactus.input steps-output/evovlerMammals.txt --inputNames okapi   --logFile steps-output/logs/preprocess-okapi.log
cactus-preprocess jobstore/10 cactus.input steps-output/evovlerMammals.txt --inputNames muskdeer   --logFile steps-output/logs/preprocess-muskdeer.log
cactus-preprocess jobstore/11 cactus.input steps-output/evovlerMammals.txt --inputNames cattle   --logFile steps-output/logs/preprocess-cattle.log
cactus-preprocess jobstore/12 cactus.input steps-output/evovlerMammals.txt --inputNames sheep   --logFile steps-output/logs/preprocess-sheep.log
cactus-preprocess jobstore/13 cactus.input steps-output/evovlerMammals.txt --inputNames goat   --logFile steps-output/logs/preprocess-goat.log
cactus-preprocess jobstore/14 cactus.input steps-output/evovlerMammals.txt --inputNames reev_muntjac   --logFile steps-output/logs/preprocess-reev_muntjac.log
cactus-preprocess jobstore/15 cactus.input steps-output/evovlerMammals.txt --inputNames ECEP   --logFile steps-output/logs/preprocess-ECEP.log
cactus-preprocess jobstore/16 cactus.input steps-output/evovlerMammals.txt --inputNames TUNLU   --logFile steps-output/logs/preprocess-TUNLU.log
cactus-preprocess jobstore/17 cactus.input steps-output/evovlerMammals.txt --inputNames HJ   --logFile steps-output/logs/preprocess-HJ.log
cactus-preprocess jobstore/18 cactus.input steps-output/evovlerMammals.txt --inputNames PZ   --logFile steps-output/logs/preprocess-PZ.log
cactus-preprocess jobstore/19 cactus.input steps-output/evovlerMammals.txt --inputNames mule_deer   --logFile steps-output/logs/preprocess-mule_deer.log
cactus-preprocess jobstore/20 cactus.input steps-output/evovlerMammals.txt --inputNames XL   --logFile steps-output/logs/preprocess-XL.log
cactus-preprocess jobstore/21 cactus.input steps-output/evovlerMammals.txt --inputNames TL   --logFile steps-output/logs/preprocess-TL.log
cactus-preprocess jobstore/22 cactus.input steps-output/evovlerMammals.txt --inputNames PL   --logFile steps-output/logs/preprocess-PL.log
cactus-preprocess jobstore/23 cactus.input steps-output/evovlerMammals.txt --inputNames mhl   --logFile steps-output/logs/preprocess-mhl.log
cactus-preprocess jobstore/24 cactus.input steps-output/evovlerMammals.txt --inputNames ML   --logFile steps-output/logs/preprocess-ML.log

## Alignment

### Round 0
cactus-blast jobstore/25 steps-output/evovlerMammals.txt steps-output/Anc07.paf --root Anc07   --logFile steps-output/logs/blast-Anc07.log
cactus-align jobstore/26 steps-output/evovlerMammals.txt steps-output/Anc07.paf steps-output/Anc07.hal --root Anc07   --logFile steps-output/logs/align-Anc07.log
hal2fasta steps-output/Anc07.hal Anc07 --hdf5InMemory > steps-output/Anc07.fa

cactus-blast jobstore/27 steps-output/evovlerMammals.txt steps-output/Anc10.paf --root Anc10   --logFile steps-output/logs/blast-Anc10.log
cactus-align jobstore/28 steps-output/evovlerMammals.txt steps-output/Anc10.paf steps-output/Anc10.hal --root Anc10   --logFile steps-output/logs/align-Anc10.log
hal2fasta steps-output/Anc10.hal Anc10 --hdf5InMemory > steps-output/Anc10.fa

cactus-blast jobstore/29 steps-output/evovlerMammals.txt steps-output/Anc16.paf --root Anc16   --logFile steps-output/logs/blast-Anc16.log
cactus-align jobstore/30 steps-output/evovlerMammals.txt steps-output/Anc16.paf steps-output/Anc16.hal --root Anc16   --logFile steps-output/logs/align-Anc16.log
hal2fasta steps-output/Anc16.hal Anc16 --hdf5InMemory > steps-output/Anc16.fa

cactus-blast jobstore/31 steps-output/evovlerMammals.txt steps-output/Anc17.paf --root Anc17   --logFile steps-output/logs/blast-Anc17.log
cactus-align jobstore/32 steps-output/evovlerMammals.txt steps-output/Anc17.paf steps-output/Anc17.hal --root Anc17   --logFile steps-output/logs/align-Anc17.log
hal2fasta steps-output/Anc17.hal Anc17 --hdf5InMemory > steps-output/Anc17.fa

cactus-blast jobstore/33 steps-output/evovlerMammals.txt steps-output/Anc19.paf --root Anc19   --logFile steps-output/logs/blast-Anc19.log
cactus-align jobstore/34 steps-output/evovlerMammals.txt steps-output/Anc19.paf steps-output/Anc19.hal --root Anc19   --logFile steps-output/logs/align-Anc19.log
hal2fasta steps-output/Anc19.hal Anc19 --hdf5InMemory > steps-output/Anc19.fa

cactus-blast jobstore/35 steps-output/evovlerMammals.txt steps-output/Anc20.paf --root Anc20   --logFile steps-output/logs/blast-Anc20.log
cactus-align jobstore/36 steps-output/evovlerMammals.txt steps-output/Anc20.paf steps-output/Anc20.hal --root Anc20   --logFile steps-output/logs/align-Anc20.log
hal2fasta steps-output/Anc20.hal Anc20 --hdf5InMemory > steps-output/Anc20.fa

cactus-blast jobstore/37 steps-output/evovlerMammals.txt steps-output/Anc23.paf --root Anc23   --logFile steps-output/logs/blast-Anc23.log
cactus-align jobstore/38 steps-output/evovlerMammals.txt steps-output/Anc23.paf steps-output/Anc23.hal --root Anc23   --logFile steps-output/logs/align-Anc23.log
hal2fasta steps-output/Anc23.hal Anc23 --hdf5InMemory > steps-output/Anc23.fa

### Round 1
cactus-blast jobstore/39 steps-output/evovlerMammals.txt steps-output/Anc05.paf --root Anc05   --logFile steps-output/logs/blast-Anc05.log
cactus-align jobstore/40 steps-output/evovlerMammals.txt steps-output/Anc05.paf steps-output/Anc05.hal --root Anc05   --logFile steps-output/logs/align-Anc05.log
hal2fasta steps-output/Anc05.hal Anc05 --hdf5InMemory > steps-output/Anc05.fa

cactus-blast jobstore/41 steps-output/evovlerMammals.txt steps-output/Anc08.paf --root Anc08   --logFile steps-output/logs/blast-Anc08.log
cactus-align jobstore/42 steps-output/evovlerMammals.txt steps-output/Anc08.paf steps-output/Anc08.hal --root Anc08   --logFile steps-output/logs/align-Anc08.log
hal2fasta steps-output/Anc08.hal Anc08 --hdf5InMemory > steps-output/Anc08.fa

cactus-blast jobstore/43 steps-output/evovlerMammals.txt steps-output/Anc13.paf --root Anc13   --logFile steps-output/logs/blast-Anc13.log
cactus-align jobstore/44 steps-output/evovlerMammals.txt steps-output/Anc13.paf steps-output/Anc13.hal --root Anc13   --logFile steps-output/logs/align-Anc13.log
hal2fasta steps-output/Anc13.hal Anc13 --hdf5InMemory > steps-output/Anc13.fa

cactus-blast jobstore/45 steps-output/evovlerMammals.txt steps-output/Anc15.paf --root Anc15   --logFile steps-output/logs/blast-Anc15.log
cactus-align jobstore/46 steps-output/evovlerMammals.txt steps-output/Anc15.paf steps-output/Anc15.hal --root Anc15   --logFile steps-output/logs/align-Anc15.log
hal2fasta steps-output/Anc15.hal Anc15 --hdf5InMemory > steps-output/Anc15.fa

cactus-blast jobstore/47 steps-output/evovlerMammals.txt steps-output/Anc22.paf --root Anc22   --logFile steps-output/logs/blast-Anc22.log
cactus-align jobstore/48 steps-output/evovlerMammals.txt steps-output/Anc22.paf steps-output/Anc22.hal --root Anc22   --logFile steps-output/logs/align-Anc22.log
hal2fasta steps-output/Anc22.hal Anc22 --hdf5InMemory > steps-output/Anc22.fa

### Round 2
cactus-blast jobstore/49 steps-output/evovlerMammals.txt steps-output/Anc11.paf --root Anc11   --logFile steps-output/logs/blast-Anc11.log
cactus-align jobstore/50 steps-output/evovlerMammals.txt steps-output/Anc11.paf steps-output/Anc11.hal --root Anc11   --logFile steps-output/logs/align-Anc11.log
hal2fasta steps-output/Anc11.hal Anc11 --hdf5InMemory > steps-output/Anc11.fa

cactus-blast jobstore/51 steps-output/evovlerMammals.txt steps-output/Anc21.paf --root Anc21   --logFile steps-output/logs/blast-Anc21.log
cactus-align jobstore/52 steps-output/evovlerMammals.txt steps-output/Anc21.paf steps-output/Anc21.hal --root Anc21   --logFile steps-output/logs/align-Anc21.log
hal2fasta steps-output/Anc21.hal Anc21 --hdf5InMemory > steps-output/Anc21.fa

### Round 3
cactus-blast jobstore/53 steps-output/evovlerMammals.txt steps-output/Anc18.paf --root Anc18   --logFile steps-output/logs/blast-Anc18.log
cactus-align jobstore/54 steps-output/evovlerMammals.txt steps-output/Anc18.paf steps-output/Anc18.hal --root Anc18   --logFile steps-output/logs/align-Anc18.log
hal2fasta steps-output/Anc18.hal Anc18 --hdf5InMemory > steps-output/Anc18.fa

### Round 4
cactus-blast jobstore/55 steps-output/evovlerMammals.txt steps-output/Anc14.paf --root Anc14   --logFile steps-output/logs/blast-Anc14.log
cactus-align jobstore/56 steps-output/evovlerMammals.txt steps-output/Anc14.paf steps-output/Anc14.hal --root Anc14   --logFile steps-output/logs/align-Anc14.log
hal2fasta steps-output/Anc14.hal Anc14 --hdf5InMemory > steps-output/Anc14.fa

### Round 5
cactus-blast jobstore/57 steps-output/evovlerMammals.txt steps-output/Anc12.paf --root Anc12   --logFile steps-output/logs/blast-Anc12.log
cactus-align jobstore/58 steps-output/evovlerMammals.txt steps-output/Anc12.paf steps-output/Anc12.hal --root Anc12   --logFile steps-output/logs/align-Anc12.log
hal2fasta steps-output/Anc12.hal Anc12 --hdf5InMemory > steps-output/Anc12.fa

### Round 6
cactus-blast jobstore/59 steps-output/evovlerMammals.txt steps-output/Anc09.paf --root Anc09   --logFile steps-output/logs/blast-Anc09.log
cactus-align jobstore/60 steps-output/evovlerMammals.txt steps-output/Anc09.paf steps-output/Anc09.hal --root Anc09   --logFile steps-output/logs/align-Anc09.log
hal2fasta steps-output/Anc09.hal Anc09 --hdf5InMemory > steps-output/Anc09.fa

### Round 7
cactus-blast jobstore/61 steps-output/evovlerMammals.txt steps-output/Anc06.paf --root Anc06   --logFile steps-output/logs/blast-Anc06.log
cactus-align jobstore/62 steps-output/evovlerMammals.txt steps-output/Anc06.paf steps-output/Anc06.hal --root Anc06   --logFile steps-output/logs/align-Anc06.log
hal2fasta steps-output/Anc06.hal Anc06 --hdf5InMemory > steps-output/Anc06.fa

### Round 8
cactus-blast jobstore/63 steps-output/evovlerMammals.txt steps-output/Anc04.paf --root Anc04   --logFile steps-output/logs/blast-Anc04.log
cactus-align jobstore/64 steps-output/evovlerMammals.txt steps-output/Anc04.paf steps-output/Anc04.hal --root Anc04   --logFile steps-output/logs/align-Anc04.log
hal2fasta steps-output/Anc04.hal Anc04 --hdf5InMemory > steps-output/Anc04.fa

### Round 9
cactus-blast jobstore/65 steps-output/evovlerMammals.txt steps-output/Anc03.paf --root Anc03   --logFile steps-output/logs/blast-Anc03.log
cactus-align jobstore/66 steps-output/evovlerMammals.txt steps-output/Anc03.paf steps-output/Anc03.hal --root Anc03   --logFile steps-output/logs/align-Anc03.log
hal2fasta steps-output/Anc03.hal Anc03 --hdf5InMemory > steps-output/Anc03.fa

### Round 10
cactus-blast jobstore/67 steps-output/evovlerMammals.txt steps-output/Anc02.paf --root Anc02   --logFile steps-output/logs/blast-Anc02.log
cactus-align jobstore/68 steps-output/evovlerMammals.txt steps-output/Anc02.paf steps-output/Anc02.hal --root Anc02   --logFile steps-output/logs/align-Anc02.log
hal2fasta steps-output/Anc02.hal Anc02 --hdf5InMemory > steps-output/Anc02.fa

### Round 11
cactus-blast jobstore/69 steps-output/evovlerMammals.txt steps-output/Anc01.paf --root Anc01   --logFile steps-output/logs/blast-Anc01.log
cactus-align jobstore/70 steps-output/evovlerMammals.txt steps-output/Anc01.paf steps-output/Anc01.hal --root Anc01   --logFile steps-output/logs/align-Anc01.log
hal2fasta steps-output/Anc01.hal Anc01 --hdf5InMemory > steps-output/Anc01.fa

### Round 12
cactus-blast jobstore/71 steps-output/evovlerMammals.txt steps-output/Anc00.paf --root Anc00   --logFile steps-output/logs/blast-Anc00.log
cactus-align jobstore/72 steps-output/evovlerMammals.txt steps-output/Anc00.paf steps-output/evolverMammals.hal --root Anc00   --logFile steps-output/logs/align-Anc00.log
hal2fasta steps-output/evolverMammals.hal Anc00 --hdf5InMemory > steps-output/Anc00.fa

## HAL merging
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc01.hal Anc01 Anc01 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc02.hal Anc02 Anc02 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc03.hal Anc03 Anc03 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc04.hal Anc04 Anc04 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc06.hal Anc06 Anc06 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc09.hal Anc09 Anc09 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc12.hal Anc12 Anc12 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc14.hal Anc14 Anc14 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc18.hal Anc18 Anc18 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc11.hal Anc11 Anc11 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc21.hal Anc21 Anc21 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc05.hal Anc05 Anc05 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc08.hal Anc08 Anc08 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc13.hal Anc13 Anc13 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc15.hal Anc15 Anc15 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc22.hal Anc22 Anc22 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc07.hal Anc07 Anc07 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc10.hal Anc10 Anc10 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc16.hal Anc16 Anc16 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc17.hal Anc17 Anc17 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc19.hal Anc19 Anc19 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc20.hal Anc20 Anc20 --merge --hdf5InMemory
halAppendSubtree steps-output/evolverMammals.hal steps-output/Anc23.hal Anc23 Anc23 --merge --hdf5InMemory

```
